**Доброго времени суток! Меня зовут Айнур!**

В данном проекте я выстыпаю аналитиком для компании, которая собирается войти в новую сферу деятельности. Для того чтобы сформировать ценностное предложение нового продукта, нужно проанализировать имеющиеся данные.

**Цель** -  Проанализировать базу данных.


**P.S.** под строками кода я буду оставлять свои комментарии и делать небольшие выводы в таком формате: **<font color="green">Комментарий:<font>** 

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas

## Ознакомление с данными

In [2]:
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**<font color="green">Комментарий:<font>** Первая таблица - `books`. В ней хранятся данные о книгах: 
 
- уникальный идентификатор книги;
    

- идентифиатор автора;
    
    
- название;
    
    
- количество страниц; 
    
    
- дата публикации; 
    
    
- идентификатор издателя.

In [3]:
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**<font color="green">Комментарий:<font>** Вторая таблица - `authors`. В ней хранятся данные об авторах:
    
    
- идентификатор автора;
    
    
- ФИО автора. 

In [4]:
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**<font color="green">Комментарий:<font>** Третья таблица - `publishers`. В ней хранятся данные об издателях:
    
    
- идентификатор издателя;
    
    
- название издательства.

In [5]:
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**<font color="green">Комментарий:<font>** Четвертая таблица - `ratings`. В ней хранятся данные о пользовательских оценках книг:
    
    
- идентификатор оценки;
    
    
- идентификатор книги;
    
        
- имя пользователя, оставившего оценку;
    
    
- оценка книги.

In [6]:
query = '''SELECT * FROM reviews '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

**<font color="green">Комментарий:<font>** Пятая таблица - `reviews`. В ней хранятся данные о пользовательских обзорах:
    
    
- идентификатор обзора;
    
    
- идентификатор книги;
    
    
- имя автора обзора;
    
    
- текст обзора.

## Задания

### Количество книг, вышедших после 1 января 2000 года

In [7]:
query = '''
SELECT count (*)
FROM books  
Where publication_date > '2000-01-01'
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**<font color="green">Комментарий:<font>** Всего у нас есть информация по **1000** книгам, **819** из них вышли после **01.01.2000** года.

### Количество обзоров и средняя оценка

Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [8]:
query = '''
select books.book_id,
       title,
       avg(ratings.rating) as avg_rating,
       count(distinct reviews.review_id) as count_of_rewiews
from books
left outer join ratings on books.book_id = ratings.book_id
left outer join reviews on books.book_id = reviews.book_id
group by books.book_id,
title
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  avg_rating  \
0          1                                       'Salem's Lot    3.666667   
1          2                 1 000 Places to See Before You Die    2.500000   
2          3  13 Little Blue Envelopes (Little Blue Envelope...    4.666667   
3          4  1491: New Revelations of the Americas Before C...    4.500000   
4          5                                               1776    4.000000   
..       ...                                                ...         ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)    3.666667   
996      997                        Xenocide (Ender's Saga  #3)    3.400000   
997      998                                    Year of Wonders    3.200000   
998      999                        You Suck (A Love Story  #2)    4.500000   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...    3.833333   

     count_of_rewiews  
0                   2  
1                   1  
2                   3  
3                   2  
4                   4  
..                ...  
995                 3  
996                 3  
997                 4  
998                 2  
999                 4  

[1000 rows x 4 columns]

**<font color="green">Комментарий:<font>** Количество обзоров и средняя оценка подсчитана!

### Издательство, выпустившее наибольшее количество книг

In [9]:
query = '''
SELECT books.publisher_id,
       publisher, 
       count(books.publisher_id) as total
FROM books  
inner join publishers on books.publisher_id = publishers.publisher_id 
Where num_pages > 50
group by books.publisher_id,
publisher
order by total desc
limit 1
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id      publisher  total
0           212  Penguin Books     42

**<font color="green">Комментарий:<font>** **Penguin Books** - издательство, которое выпустило наибольшее количество книг толще 50 страниц!

### Автор с самой высокой средней оценкой

In [10]:
query = '''
with zxc as (with wsx as (SELECT books.book_id,
       round(avg(rating), 2) as total
FROM books  
inner join ratings on books.book_id = ratings.book_id
group by books.book_id),

asd as (SELECT books.book_id,
       count(*) as ratings_counts
FROM books  
inner join ratings on books.book_id = ratings.book_id
group by books.book_id
having count(*) > 50
order by ratings_counts desc)

select *
from books 
inner join wsx on books.book_id = wsx.book_id
inner join asd on books.book_id = asd.book_id
order by total desc)

select author,
       avg(total) as avg_rating_by_ath
from zxc
inner join authors on zxc.author_id = authors.author_id
group by author
order by avg_rating_by_ath desc
limit 1
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating_by_ath
0  J.K. Rowling/Mary GrandPré              4.285

**<font color="green">Комментарий:<font>** **Джоан Роулинг** - самый рейтинговый автор!

### Пользователи, которые поставили больше 48 оценок

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [11]:
query = '''
with asd as(
with wsx as (SELECT username,
       count(*)  as count_of_rating
FROM ratings
group by username
having count(*) > 48)

select reviews.username,
       count(*) 
from reviews
inner join wsx on reviews.username = wsx.username
group by (reviews.username))

select avg(count)
from asd
'''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

**<font color="green">Комментарий:<font>** В среднем пользователи, которые поставили больше **48** оценок, оставляют **24** обзора.

## Вывод

После анализа базы данных мы можем сформировать основные ценности аудитории, а именно:

1. 819 книг вышли в тираж после 01.01.2000 года;


2. Создали таблицу, в которой отображается информация о количестве обзоров и средней оценки аудитории, например, книга «13 маленьких голубых конвертов» имеет средний рейтинг 4.6, а количество обзоров - три;


3. Penguin Books - издательство, которое выпустило наибольшее количество книг толще 50 страниц;


4. Джоан Роулинг - автор с самой высокой средней оценкой - 4.41 балла;


5. В среднем пользователи, которые поставили больше 48 оценок, оставляют 24 обзора.